httpx/http_service.go

```diff
@@ -0,0 +1,165 @@
+package httpx
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/remind101/pkg/retry"
+
+	"golang.org/x/net/context"
+)
+
+type RoundTripper interface {
+	RoundTrip(*http.Request, context.Context) (*http.Response, error)
+}
+
+type Client struct {
+	innerClient		*http.Client
+	retryTransport 	*RetryTransport
+	baseURL 		*url.URL
+}
+
+type RetryTransport struct {
+	Retrier     *retry.Retrier
+	Transport	*http.Transport
+	Client		*Client
+}
+
+func NewHTTPClient(serviceName string, baseURL *url.URL) *Client {
+	retryTransport := NewRetryTransport(serviceName)
+
+	client := &Client{
+		innerClient: &http.Client{},
+		baseURL: baseURL,
+		retryTransport: retryTransport,
+	}
+	retryTransport.Client = client
+	client.innerClient.Transport = retryTransport.Transport
+
+	return client
+}
+func NewRetryTransport(serviceName string) *RetryTransport {
+	retrier := retry.NewErrorTypeRetrier(serviceName,
+		retry.DefaultBackOffOpts,
+		(*net.OpError)(nil),
+		(*RetryableHTTPError)(nil))
+
+	// DefaultTransport but with different numerical settings
+	transport := &http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		Dial: (&net.Dialer{
+			Timeout:   15 * time.Second,
+			KeepAlive: 90 * time.Second,
+		}).Dial,
+		TLSHandshakeTimeout: 3 * time.Second,
+	}
+
+	return &RetryTransport{
+		Retrier: retrier,
+		Transport: transport,
+	}
+}
+
+func (rt *RetryTransport) RoundTrip(req *http.Request, ctx context.Context) (*http.Response, error) {
+	ret, err := rt.Retrier.Retry(func() (interface{}, error) {
+		return rt.Client.call(req, ctx)
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp, ok := ret.(*http.Response); ok {
+		return resp, nil
+	} else {
+		panic("http.Response not of correct type")
+	}
+}
+
+func (client *Client) Do(req *http.Request, ctx context.Context) (*http.Response, error) {
+	return client.retryTransport.RoundTrip(req, ctx)
+}
+
+func (client *Client) call(req *http.Request, ctx context.Context) (*http.Response, error) {
+	req.Header.Add("X-Request-Id", RequestID(ctx))
+
+	resp, err := client.innerClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode >= 500 {
+		return nil, &RetryableHTTPError{Path: req.URL.String(), StatusCode: resp.StatusCode}
+	} else if resp.StatusCode >= 300 {
+		return nil, &HTTPError{Path: req.URL.String(), StatusCode: resp.StatusCode}
+	}
+
+	return resp, nil
+}
+
+func (client *Client) ParseURL(path string) string {
+	urlString := client.UrlWithoutCreds(*client.baseURL)
+	return strings.TrimRight(urlString, "/") + path
+}
+
+// Shows the URL without information about the current username and password
+func (client *Client) UrlWithoutCreds(u url.URL) string {
+	u.User = nil
+	return u.String()
+}
+
+func (client *Client) BuildJSONRequest(method, path string, jsonData interface{}) (*http.Request, error) {
```

Comment:
Couple of things that might be improvements here:

1. This can be extracted into a function, since it doesn't actually rely on any state within Client.

```go
// NewJSONRequest generates a new http.Request with the body set to the json
// encoding of v.
func NewJSONRequest(method, path string, v interface{}) (*http.Request, error) {
        var r io.Reader
        if v != nil {
                raw, err := json.Marshal(v)
                if err != nil {
                        return nil, err
                }
                r = bytes.NewReader(raw)
        }

        req, err := http.NewRequest(method, path, r)
        if err != nil {
                return nil, err
        }
        if v != nil {
                req.Header.Set("Content-Type", "application/json")
        }
        return req, nil
}
```

2. Super minor, but the standard convention in go is to use the variable `v` for `interface{}` types.